tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,209 @@
+            The name of the data split used for evaluation in the ``dataset_dict``.
+            Default is "test".
+        seed : :class:`int`, optional
+            Used to set the RNG states at the beginning of training.
+        eval_steps : :class:`int`, optional
+            The number of steps to evaluate for. If not specified evaluation will
+            stop after a complete iteration through the ``dataloader``.
+        log_every : :class:`int`, optional
+            Log every this many steps. Default is ``1``.
+        device : ``Union[int, str, torch.device]``, optional
+            The device to evaluate on. Default to the first CUDA device if one is available,
+            otherwise CPU.
+        metric_names : ``Sequence[str]``, optional
+            The names of the metrics to track and aggregate. Default is ``("loss",)``.
+        auto_aggregate_metrics : :class:`bool`, optional
+            If ``True`` (the default), the metrics will be averaged across batches.
+            This may not be the correct behavior for some metrics (such as F1),
+            in which you should set this to ``False`` and handle the aggregation
+            internally in your model or with an :class:`EvalCallback`
+            (using :meth:`EvalCallback.post_batch()`).
+        callbacks : ``List[TrainCallback]``
+            A list of :class:`EvalCallback`.
+
+        """
+        # Set seeds.
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(seed)
+
+        check_dataset(dataset_dict, test_split)
+
+        # Resolve device.
+        _device: torch.device
+        if device is None:
+            if torch.cuda.is_available():
+                print("CUDA is available")
+                _device = torch.device("cuda")
+            else:
+                _device = torch.device("cpu")
+        elif isinstance(device, int):
+            if device >= 0:
+                _device = torch.device(f"cuda:{device}")
+            else:
+                _device = torch.device("cpu")
+        elif isinstance(device, str):
+            _device = torch.device(device)
+        elif isinstance(device, torch.device):
+            _device = device
+        else:
+            raise ValueError(f"unexpected type for 'device': '{device}'")
+        device: torch.device = _device
+
+        # Prep model.
+        model = model.eval().to(device)
+
+        # Construct dataloader.
+        dataloader: DataLoader = dataloader.construct(dataset=dataset_dict[test_split])
+
+        steps: int
+        try:
+            dataloader_len = len(dataloader)
+            steps = dataloader_len if eval_steps is None else min(dataloader_len, eval_steps)
+        except TypeError:
+            if eval_steps is None:
+                raise ConfigurationError(
+                    "You must set 'eval_steps' for streaming/iterable datasets"
```

Comment:
I really wanted to get rid of streaming datasets. They make everything more complicated, duplicate every code path, and are usually not necessary. Lazy loading solves the problem 90% of the time.

Reply:
I disagree. There are still scenarios where streaming datasets are useful. For one, your data might actually come from a stream. Another case I ran into recently is when I wanted to combine a bunch of different datasets into a multi-task mixture, but some datasets needed to be undersampled because they were too large. It was easier to implement this as a streaming dataset.

Reply:
What example do we have of datasets that are truly only available in stream form?

Mixing datasets should be easy with lazily loaded datasets. Why was it not?

Reply:
> Mixing datasets should be easy with lazily loaded datasets. Why was it not?

Take a look at https://github.com/allenai/meta-learn-prompt/blob/main/meta_learn_prompt/data/mixer_dataset.py to see how complex it got.

> What example do we have of datasets that are truly only available in stream form?

I don't have any actual examples, but I'm sure they're out there.

We already have support for streaming datasets, it really doesn't add much complexity in my opinion. So I don't understand why we'd want to get rid of it.

Reply:
I know there are plenty of downsides to streaming datasets, but my whole argument here is that someone, somewhere may want to use streaming datasets, so we might as well support them where we can especially since PyTorch and HF datasets officially support them too.

Reply:
I'm not saying we bend over backwards to support streaming datasets **everywhere**, just that we might as well support them where it's easy to. And I think it's easy to support them in `TorchEvalStep` and in `TorchTrainStep`.

Reply:
This is not the PR to address it, but streaming datasets are among the worst design choices that the PyTorch team made, and there is considerable competition for that prize. It would be much lower mental overhead if users could know that in Tango, all datasets are `DatasetDict`, and all of them support random access. It might take some code behind the scenes to realize the dream, but we are in a great place to be providing that kind of code.